silkworm/sync/sentry_client.cpp

```diff
@@ -16,137 +16,300 @@
-    auto [head_hash, head_td] = headers.head_hash_and_total_difficulty();
-    auto head_height = headers.head_height();
+    auto [head_height, head_hash, head_td] = headers.head_info();
     headers.close();
 
     log::Debug("Chain/db status", {"head hash", head_hash.to_hex()});
     log::Debug("Chain/db status", {"head td", intx::to_string(head_td)});
     log::Debug("Chain/db status", {"head height", std::to_string(head_height)});
 
-    set_status(head_hash, head_td, chain_config_);
+    set_status(head_height, head_hash, head_td, chain_config_);
 }
 
-void SentryClient::hand_shake() {
-    auto rpc = std::make_shared<rpc::HandShake>();
-    std::atomic_store(&handshake_, rpc);
-    exec_remotely(*handshake_);
-
-    SILK_TRACE << "SentryClient, hand_shake sent";
-    sentry::HandShakeReply reply = handshake_->reply();
+void SentryClient::handshake() {
+    auto supported_protocol = sync_spawn(tasks_, io_context_, sentry_client_->service()->handshake());
 
-    sentry::Protocol supported_protocol = reply.protocol();
-    if (supported_protocol < sentry::Protocol::ETH66) {
-        log::Critical("SentryClient") << "remote sentry do not support eth/66 protocol, stopping...";
+    if (supported_protocol < 66) {
+        log::Critical(kLogTitle) << "remote sentry do not support eth/66 protocol, stopping...";
         stop();
         throw SentryClientException("SentryClient exception, cause: sentry do not support eth/66 protocol");
     }
 }
 
-void SentryClient::execution_loop() {
-    using RMS = rpc::ReceiveMessages::Scope;
+static sentry::common::Message sentry_message_from_outbound_message(const OutboundMessage& outbound_message) {
+    return sentry::common::Message{
+        sentry::eth::common_message_id_from_eth_id(outbound_message.eth_message_id()),
+        outbound_message.message_data(),
+    };
+}
+
+static SentryClient::PeerIds peer_ids_from_peer_keys(const silkworm::sentry::api::api_common::Service::PeerKeys& peer_keys) {
+    SentryClient::PeerIds peer_ids;
+    for (auto& peer_key : peer_keys) {
+        peer_ids.push_back(peer_key.serialized());
+    }
+    return peer_ids;
+}
+
+SentryClient::PeerIds SentryClient::send_message_by_id(const OutboundMessage& outbound_message, const PeerId& peer_id) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_public_key = sentry::common::EccPublicKey::deserialize(peer_id);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_by_id(std::move(message), std::move(peer_public_key)));
+    return peer_ids_from_peer_keys(peer_keys);
+}
+
+SentryClient::PeerIds SentryClient::send_message_to_random_peers(const OutboundMessage& outbound_message, size_t max_peers) {
+    auto message = sentry_message_from_outbound_message(outbound_message);
+    auto peer_keys = sync_spawn(tasks_, io_context_, sentry_client_->service()->send_message_to_random_peers(std::move(message), max_peers));
```

Comment:
I need clarification on the class TaskGroup. What are its responsibilities? why it is used here?

Reply:
I will add some docs to it in a follow-up PR.